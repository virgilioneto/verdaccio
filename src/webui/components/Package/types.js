/**
 * @prettier
 * @flow
 */

export interface IProps {
  name: string;
  version: string;
  time: string;
  author: IAuthor;
  description?: string;
  keywords?: string[];
  license?: string;
}

export interface IAuthor {
  name: string;
  avatar: string;
  email: string;
}
