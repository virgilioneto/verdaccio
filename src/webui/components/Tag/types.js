/**
 * @prettier
 * @flow
 */
import type { Node } from 'react';

export interface IProps {
  children: Node;
}
