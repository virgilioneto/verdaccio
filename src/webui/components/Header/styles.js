/**
 * @prettier
 * @flow
 */

import styled, { css } from 'react-emotion';
import AppBar from '@material-ui/core/AppBar/index';
import Toolbar from '@material-ui/core/Toolbar/index';
import IconButton from '@material-ui/core/IconButton/index';

import colors from '../../utils/styles/colors';
import mq from '../../utils/styles/media';

export const NavBar = styled(AppBar)`
  && {
    background-color: ${colors.primary};
    min-height: 60px;
    display: flex;
    justify-content: center;
  }
`;

export const InnerNavBar = styled(Toolbar)`
  && {
    justify-content: space-between;
    align-items: center;
    padding: 0 15px;
    ${mq.large(css`
      padding: 0 20px;
    `)};
    ${mq.xlarge(css`
      max-width: 1240px;
      width: 100%;
      margin: 0 auto;
    `)};
  }
`;

export const Greetings = styled.span`
  && {
    margin: 0 5px 0 0;
  }
`;

export const RightSide = styled(Toolbar)`
  && {
    display: flex;
    padding: 0;
  }
`;

export const LeftSide = styled(RightSide)`
  && {
    flex: 1;
  }
`;

export const MobileNavBar = styled.div`
  && {
    align-items: center;
    display: flex;
    border-bottom: 1px solid ${colors.greyLight};
    padding: 8px;
    position: relative;
    ${mq.medium(css`
      display: none;
    `)};
  }
`;

export const InnerMobileNavBar = styled.div`
  && {
    border-radius: 4px;
    background-color: ${colors.greyLight};
    color: ${colors.white};
    width: 100%;
    padding: 0px 5px;
    margin: 0 10px 0 0;
  }
`;

export const IconSearchButton = styled(IconButton)`
  && {
    display: block;
    ${mq.medium(css`
      display: none;
    `)};
  }
`;

export const SearchWrapper = styled.div`
  && {
    display: none;
    max-width: 393px;
    width: 100%;
    display: none;
    ${mq.medium(css`
      display: flex;
    `)};
  }
`;
