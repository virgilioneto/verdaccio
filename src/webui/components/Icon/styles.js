/**
 * @prettier
 * @flow
 */

import styled, { css } from 'react-emotion';
import { IProps } from './types';

const getSize = (size: string) => {
  switch (size) {
    case 'md':
      return `
        width: 18px;
        height: 18px;
      `;
    case 'lg':
      return `
        width: 20px;
        height: 20px;
      `;
    default:
      return `
        width: 14px;
        height: 14px;
      `;
  }
};

const commonStyle = ({ size = 'sm', pointer, modifiers }: IProps) => css`
  && {
    display: inline-block;
    cursor: ${pointer ? 'pointer' : 'default'};
    ${getSize(size)};
    ${modifiers && modifiers};
  }
`;

export const SVG = styled.svg`
  && {
    ${commonStyle};
  }
`;

export const ImgWrapper = styled.span`
  && {
    ${commonStyle};
  }
`;

export const Img = styled.img`
  && {
    width: 100%;
    height: auto;
  }
`;
