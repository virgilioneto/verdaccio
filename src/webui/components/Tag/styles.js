/**
 * @prettier
 * @flow
 */

import styled from 'react-emotion';
import { ellipsis } from 'polished';

export const Wrapper = styled.span`
  && {
    vertical-align: middle;
    line-height: 22px;
    border-radius: 2px;
    color: #9f9f9f;
    background-color: hsla(0, 0%, 51%, 0.1);
    padding: 0.22rem 0.4rem;
    margin: 5px 10px 0 0;
    ${ellipsis('300px')};
  }
`;
