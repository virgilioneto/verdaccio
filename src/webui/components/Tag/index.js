/**
 * @prettier
 * @flow
 */

import React from 'react';
import type { Element } from 'react';

import { IProps } from './types';
import { Wrapper } from './styles';

const Tag = ({ children }: IProps): Element<Wrapper> => <Wrapper>{children}</Wrapper>;

export default Tag;
