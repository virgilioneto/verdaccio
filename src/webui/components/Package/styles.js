/**
 * @prettier
 * @flow
 */

import styled, { css } from 'react-emotion';
import { Link } from 'react-router-dom';
import { default as Photo } from '@material-ui/core/Avatar';
import { default as Ico } from '../Icon';
import { ellipsis } from 'polished';

import mq from '../../utils/styles/media';

import Label from '../Label';

export const Wrapper = styled.div`
  && {
    font-size: 12px;
    background-color: #fff;
    margin: 0 0 15px 0;
    transition: box-shadow 0.15s;
    box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.15);
    border-radius: 3px;
    padding: 10px;
  }
`;

// HEADER
export const Header = styled.div`
  && {
    display: flex;
    flex-direction: column;
    padding: 0 0 5px 0;
    ${mq.medium(css`
      flex-direction: row;
      justify-content: space-between;
      align-items: center;
    `)};
  }
`;

export const Name = styled.span`
  && {
    ${ellipsis('50%')};
  }
`;

export const A = styled(Link)`
  && {
    font-size: 14px;
    text-decoration: none;
    color: #3a8bff;
    font-weight: 600;
    line-height: 30px;
    flex: 1;
    padding: 0 10px 0 0;
    overflow: hidden;
    display: flex;
    align-items: center;
    :hover {
      ${Name} {
        text-decoration: underline;
      }
    }
  }
`;

export const OverviewItem = styled.span`
  && {
    display: flex;
    align-items: center;
    color: #2f273c;
    margin: 0 0 5px 0;
    ${mq.medium(css`
      margin: 0 0 0 20px;
    `)};
  }
`;

export const Overview = styled.span`
  && {
    position: relative;
    display: flex;
    flex-direction: column;
    ${mq.medium(css`
      flex-direction: row;
      ${OverviewItem} {
        :first-child {
          margin: 0;
        }
      }
    `)};
  }
`;

export const Version = styled.span`
  && {
    font-size: 12px;
    padding: 0 0 0 10px;
    margin: 0 0 0 5px;
    color: #9f9f9f;
    position: relative;
    ${ellipsis('100%')};
    :before {
      content: 'â€¢';
      position: absolute;
      left: 0;
    }
  }
`;

export const Icon = styled(Ico)`
  && {
    margin: 1px 5px 0 0;
    fill: #bfbfbf;
  }
`;

// Content
export const Field = styled.div`
  && {
    padding: 0 0 10px 0;
  }
`;

export const Content = styled.div`
  && {
    ${Field} {
      :last-child {
        padding: 0;
      }
    }
  }
`;

export const Text = styled(Label)`
  && {
    color: #908ba1;
  }
`;

export const Details = styled.span`
  && {
    margin-left: 5px;
    line-height: 14px;
    display: flex;
    flex-direction: column;
  }
`;

export const Author = styled.div`
  && {
    display: flex;
    align-items: center;
  }
`;

export const Avatar = styled(Photo)`
  && {
    width: 30px;
    height: 30px;
    background: #4b5e40;
    font-size: 15px;
  }
`;

// Footer
export const Footer = styled.div`
  && {
    padding: 5px 0 0 0;
  }
`;
