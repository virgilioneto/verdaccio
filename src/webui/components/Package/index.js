/**
 * @prettier
 * @flow
 */

import React from 'react';
import type { Element } from 'react';
import { margin } from 'polished';

import Tag from '../Tag';
import { formatDate, formatDateDistance } from '../../utils/package';

import { IProps } from './types';
import { Wrapper, Header, A, Name, Version, Overview, OverviewItem, Icon, Text, Details, Avatar, Author, Field, Content, Footer } from './styles';

const getInitialsName = (name: string) =>
  name
    .split(' ')
    .reduce((accumulator, currentValue) => accumulator.charAt(0) + currentValue.charAt(0), '')
    .toUpperCase();

const Package = ({
  name: label,
  version,
  time,
  author: { name = 'Anonymous', avatar = '', email = '' },
  description,
  license = 'Unlicensed',
  keywords,
}: IProps): Element<Wrapper> => (
  <Wrapper>
    <Header>
      <A to={`detail/${label}`}>
        <Name>{label}</Name>
        {version && <Version>{`${version} version`}</Version>}
      </A>
      <Overview>
        <OverviewItem>
          <Icon name="license" modifiers={margin('4px', '5px', '0px', '0px')} />
          {license || 'Unlicensed'}
        </OverviewItem>
        <OverviewItem>
          <Icon name="time" />
          {`Published on ${formatDate(time)} â€¢ ${formatDateDistance(time)} ago`}
        </OverviewItem>
      </Overview>
    </Header>
    <Content>
      <Field>
        <Text text="Author" modifiers={margin('0px', '0px', '5px', '0px')} />
        <Author>
          <Avatar alt={name} src={avatar}>
            {!avatar && getInitialsName(name)}
          </Avatar>
          <Details>
            <Text text={name} weight="bold" />
            {email && <Text text={email} />}
          </Details>
        </Author>
      </Field>
      {description && (
        <Field>
          <Text text="Description" modifiers={margin('0px', '0px', '5px', '0px')} />
          <span>{description}</span>
        </Field>
      )}
    </Content>
    {keywords.length > 0 && (
      <Footer>
        {keywords.map((keyword, index) => (
          <Tag key={index}>{keyword}</Tag>
        ))}
      </Footer>
    )}
  </Wrapper>
);

Package.defaultProps = {
  author: {
    name: 'Anonymous',
    avatar: '',
    email: '',
    url: '',
  },
  description: '',
  license: 'Unlicensed',
  keywords: [],
};

export default Package;
