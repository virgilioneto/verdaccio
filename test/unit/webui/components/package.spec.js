/**
 * Package component
 */

import React from 'react';
import { mount } from 'enzyme';
import Package from '../../../../src/webui/components/Package/index';
import Tag from '../../../../src/webui/components/Tag/index';
import { formatDate, formatDateDistance } from '../../../../src/webui/utils/package';
import { Version, A, Field, Details, OverviewItem } from '../../../../src/webui/components/Package/styles';

import { BrowserRouter } from 'react-router-dom';

/**
 * Generates one month back date from current time
 * @return {object} date object
 */
const dateOneMonthAgo = () => {
  const date = new Date();
  date.setMonth(date.getMonth() - 1);
  return date;
}

describe('<Package /> component', () => {
  it('should load the component', () => {
    const props = {
      name: 'verdaccio',
      version: '1.0.0',
      time: dateOneMonthAgo(),
      license: 'MIT',
      description: 'Private NPM repository',
      author: {
        name: 'Sam'
      },
      keywords: [
        "verdaccio"
      ]
    };
    const wrapper = mount(
      <BrowserRouter>
        <Package {...props} />
      </BrowserRouter>
    );

  
    // integration expectations

    // check link
    expect(wrapper.find(A).prop('to')).toEqual(`detail/${props.name}`);

    // check version
    expect(wrapper.find(Version).prop('children')).toEqual(`${props.version} version`);

    //check Author's name
    expect(wrapper.find(Details).children().first().get(0).props.children[0].props.text).toEqual(props.author.name);

    // check description
    expect(wrapper.find(Field).someWhere(n => {
      return (
        n.children().first().get(0).props.children[0].props.text === 'Description' && 
        n.children().childAt(1).containsMatchingElement(<span>{props.description}</span>)
      )
    })).toBe(true);

    // check license
    expect(wrapper.find(OverviewItem).someWhere(n => n.get(0).props.children[1] === props.license)).toBe(true);

    // check date
    expect(wrapper.find(OverviewItem).someWhere(n => 
      n.get(0).props.children[1] ===  `Published on ${formatDate(props.time)} â€¢ ${formatDateDistance(props.time)} ago`
    )).toBe(true);

    // check keyword
    expect(wrapper.find(Tag).prop('children')).toEqual(props.keywords[0]);

  });

  it('should load the component without author', () => {
    const props = {
      name: 'verdaccio',
      version: '1.0.0',
      time: dateOneMonthAgo(),
      license: 'MIT',
      author: {
        name: 'Anonymous',
        email: '',
        avatar: ''
      },
      description: 'Private NPM repository'
    };
    const wrapper = mount(
      <BrowserRouter>
        <Package {...props} />
      </BrowserRouter>
    );

    // integration expectations
    expect(wrapper.find(Details).children().first().get(0).props.children[0].props.text).toEqual('Anonymous');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
