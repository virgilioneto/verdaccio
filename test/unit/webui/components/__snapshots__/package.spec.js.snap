// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Package /> component should load the component without author 1`] = `"<div class=\\"css-f5mdyc e11fsc2k0\\"><div class=\\"css-1li2ujm e11fsc2k1\\"><a class=\\"css-xn7ljl e11fsc2k3\\" href=\\"detail/verdaccio\\"><span class=\\"css-bxt2bt e11fsc2k2\\">verdaccio</span><span class=\\"css-17xn9wj e11fsc2k6\\">1.0.0 version</span></a><span class=\\"css-x8xx5m e11fsc2k5\\"><span class=\\"css-n6zvvz e11fsc2k4\\"><svg class=\\"e11fsc2k7 css-nrrgew ej4jd2o0\\"><title>License</title><use xlink:href=\\"[object Object]#license\\"></use></svg>MIT</span><span class=\\"css-n6zvvz e11fsc2k4\\"><svg class=\\"e11fsc2k7 css-1gpgl7f ej4jd2o0\\"><title>Time</title><use xlink:href=\\"[object Object]#time\\"></use></svg>Published on 01.11.2018 â€¢ about 1 month ago</span></span></div><div class=\\"css-tywa7u e11fsc2k9\\"><div class=\\"css-r6baau e11fsc2k8\\"><div class=\\"e11fsc2k10 css-1w35jfk e1pneb170\\">Author</div><div class=\\"css-15496ft e11fsc2k12\\"><div class=\\"MuiAvatar-root-1 MuiAvatar-colorDefault-2 css-1to0t9u e11fsc2k13\\">A</div><span class=\\"css-1xj37ub e11fsc2k11\\"><div class=\\"e11fsc2k10 css-1xe0n7g e1pneb170\\">Anonymous</div></span></div></div><div class=\\"css-r6baau e11fsc2k8\\"><div class=\\"e11fsc2k10 css-1w35jfk e1pneb170\\">Description</div><span>Private NPM repository</span></div></div></div>"`;
